Relax VB Type Restrictions for If(a, b)

333fred opened the pull request:
This brings it inline with C#'s handling, and implements https://github.com/dotnet/vblang/issues/339. @AlekseyTs @cston @dotnet/roslyn-compiler for review. /cc @KathleenDollard.

333fred commented:
This builds on https://github.com/dotnet/roslyn/pull/29691, will retag reviewers when this is ready.

333fred commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -51,15 +51,30 @@ static string GetString()
 As Statement
 ");
 
-            // PROTOTYPE(null-operator-enhancements): lines 8 and 9 appear to be entirely redundant.
```
This is fallout from https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs#L124, which unconditionally stores the lhs of a `x ?? y` expression, even if the lhs is a local and does not need to be stored.

333fred commented:
@cston @AlekseyTs @dotnet/roslyn-compiler this is ready for review.

333fred commented:
@dotnet-bot retest windows_debug_vs-integration_prtest please

jcouv commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb:
```diff
@@ -2096,9 +2096,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-            If Not hasErrors AndAlso Not (boundFirstArg.IsNothingLiteral OrElse boundFirstArg.Type.IsNullableType OrElse boundFirstArg.Type.IsReferenceType) Then
+            '  If there are still no errors check the original type of the first argument. If it's
+            '  a non-nullable value type, generate IllegalCondTypeInIIF
+            If Not hasErrors AndAlso Not boundFirstArg.IsNothingLiteral AndAlso boundFirstArg.Type.IsValueType AndAlso Not boundFirstArg.Type.IsNullableType() Then
```
Should this new behavior be gated by language version and be documented in some list of VB16 features?
(I feel it should, since there is an accompanying spec change)

333fred replied:
Good catch, we need to do this for C# as well.

---
In reply to: [216848436](https://github.com/dotnet/roslyn/pull/29724#discussion_r216848436) [](ancestors = 216848436)

333fred commented:
@dotnet-bot test this please

333fred commented:
@dotnet-bot retest microbuild_prtest please

333fred commented:
@dotnet-bot retest windows_coreclr_release_prtest please

JohnTortugo commented:
@dotnet-bot retest windows_coreclr_release_prtest please

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -5366,4 +5366,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>To use a variable of an unconstrained type parameter in a null-coalescing assignment, please use language version {0} or greater.</value>
+  </data>
+  <data name="IDS_FeatureUnconstrainedTypeParameterInNullCoalescingOperator" xml:space="preserve">
+    <value>unconstrained type parameters in null coalescing operator</value>
```
>null coalescing operator [](start = 44, length = 24)

It looks like the language specification calls this: "Conditional Expression"

AlekseyTs commented on src/Compilers/VisualBasic/Portable/VBResources.resx:
```diff
@@ -5533,4 +5533,7 @@
     <value>interpolated strings</value>
   </data>
+  <data name="FEATURE_UnconstrainedTypeParameterInCondtional" xml:space="preserve">
+    <value>unconstrained type parameter variables in two operand conditionals</value>
```
>conditionals [](start = 65, length = 12)

It looks like the language specification calls this: "Conditional Expression"

AlekseyTs replied:
>variables [](start = 40, length = 9)

Why variables?

AlekseyTs commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1583,6 +1583,7 @@ internal enum ErrorCode
         ERR_FeatureNotAvailableInVersion8 = 8400,
         ERR_AltInterpolatedVerbatimStringsNotAvailable = 8401,
         WRN_DefaultLiteralConvertedToNullIsNotIntended = 8402,
+        ERR_UnconstrainedTypeParameterInNullCoalescingNotAvailable = 8403,
```
>ERR_UnconstrainedTypeParameterInNullCoalescingNotAvailable [](start = 8, length = 58)

It looks like this error code is not used and that matches what I would expect. Remove?

cston commented on src/Compilers/VisualBasic/Portable/Errors/Errors.vb:
```diff
@@ -2018,5 +2018,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         FEATURE_LeadingDigitSeparator
         FEATURE_PrivateProtected
         FEATURE_InterpolatedStrings
+        FEATURE_UnconstrainedTypeParameterInCondtional
```
>Condtional [](start = 44, length = 10)

Typo

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3381,10 +3381,23 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
+                {
+                    if (!CheckFeatureAvailability(node, MessageID.IDS_FeatureUnconstrainedTypeParameterInNullCoalescingOperator, diagnostics))
+                    {
+                        return new BoundNullCoalescingOperator(node, leftOperand, rightOperand, Conversion.NoConversion, CreateErrorType());
```
>return new BoundNullCoalescingOperator(node, leftOperand, rightOperand, Conversion.NoConversion, CreateErrorType()); [](start = 24, length = 116)

We should report diagnostics and keep binding normally. In general, language version shouldn't affect how we parse\bind things apart from diagnostics.

333fred replied:
I'm attempting to replicate what happened prior to C# 8.0 for unconstrained type parameters, which was to simply report an error on this node and stop binding.

---
In reply to: [217204494](https://github.com/dotnet/roslyn/pull/29724#discussion_r217204494) [](ancestors = 217204494)

AlekseyTs replied:
> I'm attempting to replicate what happened prior to C# 8.0 for unconstrained type parameters, which was to simply report an error on this node and stop binding.

We don't do this. Language version should not have an effect like this. The only effect it can have is to report diagnostics

---
In reply to: [217216299](https://github.com/dotnet/roslyn/pull/29724#discussion_r217216299) [](ancestors = 217216299,217204494)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder.vb:
```diff
@@ -1067,6 +1067,19 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+        Private Shared Function CheckFeatureAvailability(feature As InternalSyntax.Feature, node As SyntaxNode, diagnostics As DiagnosticBag) As Boolean
+            Dim langVersion As LanguageVersion = DirectCast(node.SyntaxTree.Options, VisualBasicParseOptions).LanguageVersion
+            Dim requiredVersion As LanguageVersion = InternalSyntax.FeatureExtensions.GetLanguageVersion(feature)
```
>Dim requiredVersion As LanguageVersion = InternalSyntax.FeatureExtensions.GetLanguageVersion(feature) [](start = 12, length = 101)

It looks like from this line down the code duplicates what ```Friend Shared Function CheckFeatureAvailability(diagnostics As DiagnosticBag, location As Location, languageVersion As LanguageVersion, feature As Feature) As Boolean``` in the Parser does.

333fred replied:
It does, because I didn't want to make the Binder depend on a static method in the Parser. This is what C# does.

---
In reply to: [217210173](https://github.com/dotnet/roslyn/pull/29724#discussion_r217210173) [](ancestors = 217210173)

AlekseyTs replied:
>  I didn't want to make the Binder depend on a static method in the Parser

This is where we keep the helper and we are using it elsewhere as well. Please delegate to it instead of duplicating the code

---
In reply to: [217217240](https://github.com/dotnet/roslyn/pull/29724#discussion_r217217240) [](ancestors = 217217240,217210173)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb:
```diff
@@ -2096,11 +2096,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                ' ensure that the feature is enabled and report an error if it is not
+                If Not boundFirstArg.Type.IsValueType AndAlso Not boundFirstArg.Type.IsNullableType() Then
+                    If Not CheckFeatureAvailability(InternalSyntax.Feature.UnconstrainedTypeParameterInConditional, node, diagnostics) Then
+                        hasErrors = True
```
>hasErrors = True [](start = 24, length = 16)

Same as for C#, I don't think this should be done

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb:
```diff
@@ -2096,11 +2096,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
-            '  if there are still no errors check the original type of the first argument to be 
-            '       of a reference or nullable type, generate IllegalCondTypeInIIF otherwise
+            '  If there are still no errors check the original type of the first argument. First, we check
```
>'  If there are still [](start = 12, length = 21)

It looks like an alignment is off

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3381,10 +3381,23 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
-                return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
+                // Prior to C# 8.0, the spec said that the left type must be either a reference type or a nullable value type. This was relaxed
+                // with C# 8.0, so if the feature is not enabled then issue a diagnostic and return
+                if (!optLeftType.IsValueType && !isLeftNullable)
```
> && !isLeftNullable [](start = 44, length = 19)

Is this condition redundant?

AlekseyTs replied:
> Same answer as in VB. This is not redundant as unconstrained type parameters will hit this.

The question is about ``` && !isLeftNullable``` part

---
In reply to: [217216067](https://github.com/dotnet/roslyn/pull/29724#discussion_r217216067) [](ancestors = 217216067,217212991)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb:
```diff
@@ -2096,11 +2096,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                hasErrors = True
+                ' VB 16 changed the requirements on the first operand to permit unconstrained type parameters. If we're in that scenario,
+                ' ensure that the feature is enabled and report an error if it is not
+                If Not boundFirstArg.Type.IsValueType AndAlso Not boundFirstArg.Type.IsNullableType() Then
```
> AndAlso Not boundFirstArg.Type.IsNullableType() [](start = 53, length = 48)

Is this condition redundant?

333fred replied:
No. Unconstrained type parameters will hit this, and depending on the language version will bind successfully or not.

---
In reply to: [217213079](https://github.com/dotnet/roslyn/pull/29724#discussion_r217213079) [](ancestors = 217213079)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/CodeGen/EmitExpression.vb:
```diff
@@ -1304,7 +1304,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGen
             Debug.Assert(expr.ConvertedTestExpression Is Nothing, "coalesce with nontrivial test conversions are lowered into ternary.")
             Debug.Assert(expr.Type = expr.ElseExpression.Type)
-            Debug.Assert(expr.Type.IsReferenceType)
+            Debug.Assert(expr.Type.IsReferenceType OrElse Not expr.Type.IsValueType)
```
>expr.Type.IsReferenceType OrElse  [](start = 25, length = 33)

Remove the first condition?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb:
```diff
@@ -200,7 +200,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
-            Debug.Assert(testExpr.Type.IsReferenceType)
+            ' At this point, we must either have a reference type, or an unconstrained type parameter (which is neither reference nor value)
+            Debug.Assert(testExpr.Type.IsReferenceType OrElse Not testExpr.Type.IsValueType)
```
>testExpr.Type.IsReferenceType OrElse  [](start = 25, length = 37)

Remove the first condition?

JohnTortugo commented:
@dotnet-bot retest microbuild_prtest please

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb:
```diff
@@ -181,8 +181,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-            ' Test expression may only be of a reference or nullable type
-            Debug.Assert(testExpr.IsNothingLiteral OrElse testExpr.Type.IsReferenceType OrElse testExpr.Type.IsNullableType)
+            ' Test expression may not be a non-nullable value type
+            Debug.Assert(testExpr.IsNothingLiteral OrElse Not (testExpr.Type.IsValueType AndAlso Not testExpr.Type.IsNullableType()))
```
>testExpr.Type.IsNullableType() [](start = 101, length = 30)

Can this condition ever be true?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenIfOperator.vb:
```diff
@@ -1247,5 +1247,105 @@ Friend Module BIFOpResult0011mod
+    End Module
+    </file>
+</compilation>, parseOptions:=TestOptions.Regular15_5).AssertTheseDiagnostics(<![CDATA[
+BC36716: Visual Basic 15.5 does not support unconstrained type parameter variables in two operand conditionals.
```
No information which version does?

333fred replied:
BC36716 doesn't have this information for any feature that uses it, going back to VB 9.0. We can add it if we want to, but that is out of scope for this PR.

---
In reply to: [217216426](https://github.com/dotnet/roslyn/pull/29724#discussion_r217216426) [](ancestors = 217216426)

AlekseyTs replied:
> We can add it if we want to, but that is out of scope for this PR.

Could you open an issue to follow up?

---
In reply to: [217221037](https://github.com/dotnet/roslyn/pull/29724#discussion_r217221037) [](ancestors = 217221037,217216426)

333fred replied:
Created [https://github.com/dotnet/roslyn/issues/29874](https://github.com/dotnet/roslyn/issues/29874)

---
In reply to: [217559319](https://github.com/dotnet/roslyn/pull/29724#discussion_r217559319) [](ancestors = 217559319,217221037,217216426)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenIfOperator.vb:
```diff
@@ -1247,5 +1247,105 @@ Friend Module BIFOpResult0011mod
+    <file name="a.vb">
+Friend Module Mod1
+        Sub M1(Of T)(arg1 As T, arg2 As T)
+            System.Console.WriteLine(If(arg1, arg2))
```
>If(arg1, arg2) [](start = 37, length = 14)

It would be good to have a test for a scenario when there is a conversion for the test expression

333fred replied:
Thanks, discovered an issue in Emit with this test.

---
In reply to: [217217442](https://github.com/dotnet/roslyn/pull/29724#discussion_r217217442) [](ancestors = 217217442)

AlekseyTs replied:
> discovered an issue in Emit with this test.

Do we have similar test for C#?

---
In reply to: [217544873](https://github.com/dotnet/roslyn/pull/29724#discussion_r217544873) [](ancestors = 217544873,217217442)

333fred replied:
Yes, [https://github.com/dotnet/roslyn/blob/features/null-operator-enhancements/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs#L1388](https://github.com/dotnet/roslyn/blob/features/null-operator-enhancements/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs#L1388)

---
In reply to: [217546650](https://github.com/dotnet/roslyn/pull/29724#discussion_r217546650) [](ancestors = 217546650,217544873,217217442)

AlekseyTs commented:
Done with review pass (iteration 6)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [420831568](https://github.com/dotnet/roslyn/pull/29724#issuecomment-420831568) [](ancestors = 420831568)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundTernaryConditionalExpression.vb:
```diff
@@ -18,7 +18,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     WhenFalse.AssertRValue()
                 End If
-                Debug.Assert(Condition.IsNothingLiteral() OrElse Condition.Type.IsBooleanType() OrElse Condition.Type.IsReferenceType())
+                Debug.Assert(Condition.IsNothingLiteral() OrElse Condition.Type.IsBooleanType() OrElse Condition.Type.IsReferenceType() OrElse Not Condition.Type.IsValueType)
```
>OrElse Condition.Type.IsReferenceType() [](start = 96, length = 39)

It looks like the ```OrElse Condition.Type.IsReferenceType()``` can be removed.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/CodeGen/EmitStatement.vb:
```diff
@@ -800,7 +800,8 @@ OtherExpressions:
 
                     Dim conditionType = condition.Type
-                    If conditionType.IsReferenceType AndAlso Not IsVerifierReference(conditionType) Then
+                    If (conditionType.IsReferenceType OrElse Not conditionType.IsValueType) AndAlso
```
>conditionType.IsReferenceType OrElse [](start = 24, length = 36)

Remove ```conditionType.IsReferenceType OrElse``` part?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenIfOperator.vb:
```diff
@@ -1247,5 +1247,125 @@ Friend Module BIFOpResult0011mod
+        System.Console.WriteLine(If(arg1, arg2))
+    End Sub
+
+    Sub M2(Of T1, T2 As T1)(arg1 as T1, arg2 As T2)
```
>Sub M2(Of T1, T2 As T1)(arg1 as T1, arg2 As T2) [](start = 4, length = 47)

Consider verifying behavior:
- int, int
- object (null/not-null), int

Same for the corresponding C# test. Also, it doesn't look like C# test verifies IL

AlekseyTs replied:
> - object (null/not-null), int

I guess this is not interesting, this one is - object, int? (null/not-null)

---
In reply to: [217561188](https://github.com/dotnet/roslyn/pull/29724#discussion_r217561188) [](ancestors = 217561188)

AlekseyTs commented:
Done with review pass (iteration 7)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [421181700](https://github.com/dotnet/roslyn/pull/29724#issuecomment-421181700) [](ancestors = 421181700)

333fred commented:
@dotnet-bot retest windows_release_unit32_prtest please

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3381,10 +3381,20 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
             {
-                return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
+                // Prior to C# 8.0, the spec said that the left type must be either a reference type or a nullable value type. This was relaxed
+                // with C# 8.0, so if the feature is not enabled then issue a diagnostic and return
```
Minor point: The second sentence seems unnecessary.

cston commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb:
```diff
@@ -2096,11 +2096,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                ReportDiagnostic(diagnostics, node.FirstExpression, ERRID.ERR_IllegalCondTypeInIIF)
-                hasErrors = True
+                ' VB 16 changed the requirements on the first operand to permit unconstrained type parameters. If we're in that scenario,
+                ' ensure that the feature is enabled and report an error if it is not
```
Second sentence seems unnecessary.

jcouv commented:
@333fred It would be good to have a record of this feature.
Can you create a VB lang issue (if there isn't one already) and let Kathleen know that it should be marked as 16.0 "candidate"? (or some VB LDM members who have permission on VB lang)
Thanks

333fred commented:
@jcouv https://github.com/dotnet/vblang/issues/339 :)

KathleenDollard commented:
Done
________________________________
From: Fred Silberberg <notifications@github.com>
Sent: Thursday, September 20, 2018 8:45 AM
To: dotnet/roslyn
Cc: Kathleen Dollard; Mention
Subject: Re: [dotnet/roslyn] Relax VB Type Restrictions for If(a, b) (#29724)


@jcouv<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjcouv&data=02%7C01%7Ckathleen.dollard%40microsoft.com%7Ccf595ae364064294986308d61f1024b3%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636730551529255737&sdata=6SQ8rUbrtsfnPVI5oqNXWIRh0HVCcBcBCEcU02kL8SI%3D&reserved=0> dotnet/vblang#339<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fvblang%2Fissues%2F339&data=02%7C01%7Ckathleen.dollard%40microsoft.com%7Ccf595ae364064294986308d61f1024b3%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636730551529265742&sdata=8DxefWRQx2RTug6zlRUwF0nXoqviVbjSYK5%2B4WQFYcw%3D&reserved=0> :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Froslyn%2Fpull%2F29724%23issuecomment-423232029&data=02%7C01%7Ckathleen.dollard%40microsoft.com%7Ccf595ae364064294986308d61f1024b3%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636730551529265742&sdata=Jkd8VU0LnmG8cPwoqXhf5BBIhiyCXoLX60rAEPiODiE%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFktXiObeKCBQR7-ZLszlbJH0A9HH7lWks5uc7gugaJpZM4WfQoL&data=02%7C01%7Ckathleen.dollard%40microsoft.com%7Ccf595ae364064294986308d61f1024b3%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636730551529275751&sdata=hkmLgS2JAa%2F12%2Be%2F%2B8UApFdEbUZORK8lZK9cyNK9bzs%3D&reserved=0>.
